<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html><head><title>LDIF - Linked Data Integration Framework</title>





<style>
body { background: white; color: black; font-family: sans-serif; line-height: 1.4em; padding: 2.5em 3em; margin: 0; }
:link { color: #00c; }
:visited { color: #609; }
a:link img { border: none; }
a:visited img { border: none; }
h1, h2, h3 { background: white; color: #800; }
h1 { font: 170% sans-serif; margin: 0; }
h2 { clear: both; font: 140% sans-serif; margin: 2.5em 0 -0.5em 0; }
h3 { font: 120% sans-serif; margin: 1.5em 0 -0.5em 0; }
h4 { font: bold 100% sans-serif; }
h5 { font: italic 100% sans-serif; }
h6 { font: small-caps 100% sans-serif; }
.hide { display: none; }
pre { background: #fff6bb; font-family: monospace; line-height: 1.2em; padding: 1em 2em; }
dt { font-weight: bold; margin-top: 0; margin-bottom: 0; }
dd { margin-top: 0; margin-bottom: 0; }
code, tt { font-family: monospace; }
ul.toc { list-style-type: none; }
ol.toc li a { text-decoration: none; }
.note { color: red; }
#header { border-bottom: 1px solid #ccc; }
#logo { float: right; }
#authors { clear: right; float: right; font-size: 80%; text-align: right; }
#content { clear: both; margin: 2em auto 0 0; text-align: justify }
#download, #demo { float: left; font-family: sans-serif; margin: 1em 0 1.5em; text-align: center; width: 50%; }
#download h2, #demo h2 { font-size: 125%; margin: 1.5em 0 -0.2em 0; }
#download small, #demo small { color: #888; font-size: 80%; }
#footer { border-top: 1px solid #ccc; color: #aaa; margin: 2em 0 0; }
@media Print {
* { font-size: 92%; }
body { padding: 0; line-height: 1.2em; }
#content { margin: 0; width: 100%; }
}
@media Aural {
h1 { stress: 20; richness: 90; }
h2 { stress: 20; richness: 90; }
h3 { stress: 20; richness: 90; }
.hide { speak: none; }
dt { pause-before: 20%; }
pre { speak-punctuation: code; }
}
.Stil1 {color: #FF0000}
table {border-collapse: collapse; margin-bottom: .75em; border-top: 1px solid #D4DCE8; border-bottom: 1px solid #D4DCE8}
table th {text-align: left; background: #EAF3FA; padding: .5em .9em; height: 21px; font-weight: bold; font-size: 1em; color: #333}
table th a {text-decoration: none !important}
table td {vertical-align: top; text-align: left; padding: .5em 1em; border-bottom: 1px solid #eee}
table tr.even {background-color: #fff}
table tr.odd {background-color: #f6f6f6}
</style>

<meta http-equiv="Content-Type" content="text/html; charset=UTF-8"></head><body>
<div id="logo" align="right"><a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/index.html"><img src="img/fu-logo.gif" alt="Freie UniversitÃ¤t Berlin Logo" border="0"></a>&nbsp;<br>
  <a href="http://mediaeventservices.com"><img src="img/logo.gif" alt="Media Event Services"></a></div>

<div id="header"> <h1 style="font-size: 250%;">LDIF - Linked Data Integration Framework </h1></div>

<div id="tagline">A component for building Linked Data applications
</div>
<div id="authors"><a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/team/SchultzAndreas.html">Andreas
Schultz</a><br>
<a href="mailto:andrea.matteini@gmail.com">Andrea Matteini</a><br>
<a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/team/IseleRobert.html">Robert
Isele</a><br>
<a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/team/BizerChristian.html">Chris
Bizer</a><br>
<a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/team/BeckerChristian.html">Christian Becker</a><br>
</div>
<div id="content">
<p>The LDIF - Framework translates heterogeneous Linked Data from the Web into <br>
a clean, local target representation while keeping track of data
provenance.</p>
<div id="download">
<h2><a href="http://www4.wiwiss.fu-berlin.de/bizer/ldif/releases/ldif-0.2.zip">Download LDIF</a><br>
</h2>
<small>v0.2, released 2011-08-25</small>
</div>
<h2 id="news">News</h2>
<ul>
 <li><strong>8/25/2011: Version 0.2 released.</strong> The second
LDIF release provides improved performance (faster data loading,
parallelization of the data translation), smaller memory footprint, a
new N-Triples output module, new performance evaluation results for use
cases up to 100 million triples.<br>
</li>
 <li><strong>6/29/2011: Version 0.1 released.</strong> This alpha
version provides for translating data that is represented using
different source vocabularies into a single target vocabulary and for
replacing different URIs that refer to the same real-world entity with
a single target URI.</li>
</ul>
<h2 id="contents">Contents</h2>
<ol class="toc">
  <li><a href="#about">About LDIF</a></li>
  <li><a href="#components">LDIF components</a></li>
  <li><a href="#quickstart">Quick start</a></li>
  <li><a href="#examples">Example: Using LDIF for integrating Life
Science Data</a></li>
  <li><a href="#performance">Performance Evaluation</a> </li>
  <li><a href="#config">Configuration options</a></li>
  <li><a href="#code">Source code and development</a></li>
  <li><a href="#history">Version history</a></li>
  <li><a href="#feedback">Support and Feedback</a></li>
  <li><a href="#related">References</a></li>
</ol>
<h2 id="about"><a name="about"></a>1. About LDIF<br>
</h2>
<p>The <a href="http://linkeddatabook.com/editions/1.0/#htoc23">Web of
Linked Data</a>
grows rapidly and contains data from a wide range of different domains,
including life science data, geographic data, government data, library
and media data, as well as cross-domain datasets such as DBpedia or
Freebase. <a href="http://linkeddatabook.com/editions/1.0/#htoc75">Linked
Data applications</a> that want to consume data from this global data
space face the challenges that:</p>
<ol>
  <li> data sources use a wide range of different RDF vocabularies to
represent data about the same type of entity. </li>
  <li>the
same real-world entity, for instance a person or a place, is
identified with different URIs within different data sources. </li>
</ol>
<p>This usage of different vocabularies as well as the usage of URI
aliases makes it very cumbersome for an application
developer to write <a href="http://www.w3.org/TR/rdf-sparql-query/">SPARQL</a>
queries against Web data which originates from multiple sources. In
order to ease using Web data in the application context, it is thus
advisable to translate data to a single target vocabulary (<a href="http://linkeddatabook.com/editions/1.0/#htoc86">vocabulary
mapping</a>) and to replace URI aliases with a single target URI on the client
side (<a href="http://linkeddatabook.com/editions/1.0/#htoc87">identity
resolution</a>), before starting to ask SPARQL queries against the
data. </p>
<p>Up-till-now, there have not been any integrated tools that help
application developers with these tasks. With LDIF, we try to fill this
gap and provide an an open-source Linked Data Integration Framework
that can be used by Linked Data applications to translate Web data and
normalize URI while keeping track of data provenance.<br>
<br>
LDIF provides an expressive mapping language for translating data from
the various vocabularies that are used on the Web into a consistent,
local target vocabulary. LDIF includes an identity resolution component
which discovers URI aliases in the input data and replaces them with a
single target URI based on user-provided matching heuristics. For
provenance tracking, the LDIF framework employs the Named Graphs data
model.<br>
 </p>
<p>The figure below shows the schematic <a href="http://linkeddatabook.com/editions/1.0/#htoc84">architecture of
Linked Data applications</a>
that
implement the crawling/data warehousing pattern. The figure highlights
the steps of the data integration process that are currently
supported by LDIF.<br>
<br>
<span class="Stil1"> </span> </p>
<p style="text-align: left; margin-left: 120px;"><img style="border: 1px solid ; width: 650px; height: 501px;" alt="Example-architecture of an integration aware Linked Data application" src="img/ldarchitecture.png"></p>
<p style="text-align: left; margin-left: 120px;">&nbsp;</p>
<h2><a name="components"></a>2. LDIF Components<br>
</h2>
<p>The LDIF Framework consists of the Runtime Environment and a set of
pluggable modules. The pluggable modules are organized as data access
components, a data transformation components and a data output components.
</p>
<p><img src="img/LDIFcomponents.png" height="92" width="299"></p>
Currently, we have implemented the following modules: <br>
<h3>Data Access: N-Quads Loader<br>
</h3>
<p>The current version of LDIF expects input data to be represented as Named
Graphs and be stored in <a href="http://sw.deri.org/2008/07/n-quads/">N-Quads
format</a> accessible by a URL. The graph URI is used for
provenance tracking. Provenance meta-information describing the graphs
can be provided within a specific provenance graph. The name of
this provenance graph can be set in the configuration file. LDIF does
not make any assumptions about the provenance vocabulary that is used
to describe the graphs, meaning that you can use your provenance
vocabulary of choice. Currently, the provenance information is just
copied to the final output. In future releases, we will use the
provenance information for data quality assessment and data fusion (see
<a href="#nextSteps">Next Steps</a>).<br>
</p>
<h3>Transformation: R2R Data Translation
</h3>
<p>LDIF employs the <a href="http://www4.wiwiss.fu-berlin.de/bizer/r2r/">R2R Framework</a>
to translate Web data that is represented using terms from different
vocabularies into a single target vocabulary. Vocabulary mappings are
expressed using the <a href="http://www4.wiwiss.fu-berlin.de/bizer/r2r/spec/">R2R Mapping
Language</a>.
The language provides for simple transformations as well as for more
complex structural transformations and property value transformations
such as normalizing different units of measurement or complex string
manipulations. The syntax of the
R2R Mapping Language is very similar to the query language SPARQL,
which eases the learning curve. The expressivity of the language
enabled us to deal with all requirements that we have encountered so
far when translatingLinked Data from the Web into a target representation (evaluation in [2]).  </p>
<p>An overview and examples for mappings are given on the <a href="http://www4.wiwiss.fu-berlin.de/bizer/r2r/">R2R website</a>.<br>
The <a href="http://www4.wiwiss.fu-berlin.de/bizer/r2r/spec/">specification
and user manual</a> is provided as a separate document.<br>
</p>
<h3>Transformation: Silk Identity Resolution </h3>
<p>
LDIF employs the <a href="http://www4.wiwiss.fu-berlin.de/bizer/silk/">Silk
Link Discovery Framework</a>
to find different URIs that are used within different data sources to
identify the same real-world entity. For each set of duplicates which
have been identified by Silk, LDIF replaces all URI aliases with a
single target URI within the output data. In addition, it adds
owl:sameAs links pointing at the original URIs, which makes it possible
for applications to refer back to the data sources on the Web. If the
LDIF input data already contains owl:sameAs links, the referenced URIsare normalized accordingly (optional, see  <a href="#config">configuration</a>). Silk is a
flexible identity
resolution framework that allows the user to specify identity
resolution heuristics which combine different types of matchers using
the declarative <a href="http://www4.wiwiss.fu-berlin.de/bizer/silk/spec/">Silk - Link
Specification Language</a>.</p>
<p>An overview and examples can be found on the <a href="http://www4.wiwiss.fu-berlin.de/bizer/silk/">Silk website</a>.<br>
</p>
<h3 style="text-align: left;">Data Output: N-Quads Writer<br>
</h3>
<p>The N-Quads writer dumps the final output of the integration
workflow into a single N-Quads file. This file contains the translated
versions of all graphs from the input graph set as well as the content
of the provenance graph and sameAs-links. <br>
</p>
<h3 style="text-align: left;">Data Output: N-Triples Writer<br>
</h3>
<p>The N-Triples writer dumps the final output of the integration
workflow into a single N-Triples file.<br>
</p>
<h3 style="text-align: left;">Runtime Environment </h3>
<p style="text-align: left;"> The Runtime Environment manages the data
flow between the various stages and the caching of the intermediate
results. In order to parallelize the data processing, the data is
partitioned into entities prior to supplying it to a transformation
module. An entity represents a Web resource together with all data that
is required by a transformation module to process this resource.
Entities consist of one or more graph paths and include a provenance
URI for each node. Each transformation module specifies which paths
should be included into the entities it processes. Splitting the work
into fine-granular entities, allows LDIF to parallelize the work into
multiple threats and will, in the next release, allow thework to be
parallelized on multiple machines using Hadoop. For the LDIF V0.2
release, we have implemented only an
in-memory version of the runtime environment. This implementation keeps
all intermediate results in
memory. It is very fast but scales only depending on the amount of
memory
available. For instance, integrating 25 million triples required 5 GB
memory within one of our <a href="benchmark.html">experiments</a>. </p>
<h3><a name="nextSteps"></a>Next steps for LDIF</h3>
<p>LDIF V0.2 provides data translation and identity resolution
functionality. Over the next months, we plan to extend LDIF along the
following lines:<br>
</p><ol>
  
  <li>Add <strong>Web Data Access Modules</strong> (Linked Data
  Crawler, SPARQL Endpoint Reader, Remote RDF File Loader) as well as a
  scheduling component which provides for regularly updating the local
  input data cache.</li>
  <li>Implement a <strong>Hadoop Version of the Runtime Environment </strong>in
order to be able to scale to really large amounts of input data.
Processes and data will be distributed over a cluster of machines.</li>

  
  <li>Add a <strong>Data Quality Evaluation and Data Fusion Module</strong>
which allows Web data to be filtered according to different data
quality assessment policies and provides for fusing Web data according
to different conflict resolution methods.</li>
  <li><strong>Flexible integration workflow</strong>. Currently the
integration flow is static and can only be influenced by predefined
configuration parameters. We plan to make the workflow and its
configuration more flexible in order to make it easier to include
additional modules that cover other data integration aspects. </li>
  </ol>

<h2 id="quickstart"><a name="quickstart"></a>3. Quick start</h2>
<br>
In order to experiment with LDIF follow these steps:<br>
<ol>
  <li>Download the latest release (download link at the top of this
page)</li>
  <li>Unpack the archive to an arbitrary location</li>
  <li>To run the command line tool type the following in the root
directory of LDIF:<br>
  </li>
</ol>
<pre>bin/run &lt;path-to-ldif-config-file&gt;<br><br>bin\run.bat &lt;path-to-ldif-config-file&gt;<br></pre>
For details about the configuration file and parameters, see <a href="#config">Section 6</a>. A detailed example is given next.<br>

<h2 id="examples"><a name="examples"></a>4. Example: Using LDIF to
integrate Life Science Data </h2>
<br>
<p>This example shows how LDIF is applied to integrate data originating from five Life Science  sources. </p>
<p>The example is taken from a joined project with <a href="http://www.vulcan.com/TemplateHome.aspx?contentId=1">Vulcan Inc</a>. and <a href="http://www.ontoprise.de/">ontoprise GmbH</a> about extending <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/BeckerBizerErdmannGreaves-SMW-LDE-Poster-ISWC2010.pdf">Semantic Media Wiki+ with a Linked Data Integration Framework</a>. <br>
  <br>
  In this example, the following data sources are translated into a
  common <a href="resources/Wiki.owl">Wiki ontology</a>:</p>
<ul>
  <li><a href="http://mouse.brain-map.org/">Allen Mouse Brain Atlas</a>  is a growing
      collection of online public resources integrating extensive gene
      expression and neuroanatomical data. </li>
  <li><a href="http://www.genome.jp/kegg/genes.html">KEGG GENES</a>,
        a collection of gene catalogs for all complete genomes generated from
        publicly available resources, mostly NCBI RefSeq</li>
  <li><a href="http://www.genome.jp/kegg/pathway.html">KEGG Pathway</a>,
            a collection of pathway maps representing knowledge on the molecular
          interaction and reaction networks</li>
  <li><a href="http://pharmgkb.org/">PharmGKB</a>, which provides
          data on gene information, disease and drug pathways, and SNP variants</li>
  <li><a href="http://www.uniprot.org/">Uniprot</a>, which provides
          information on protein sequence and function </li>
  </ul>
<p>The following mapping file provides for translating the vocabularies used by the source datasets into the Wiki ontology. </p>
<ul>
      <li>R2R mapping file: <a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/mappings/ALL-to-Wiki.r2r.ttl?rev=176428845b9594e28a2f0362916de23cc821502c">ALL-toWiki.r2r.ttl</a>
        <span style="text-decoration: underline;"></span></li>
    </ul>
<p>The following Silk identity resolution heuristics are used to find
genes and other expressions that are described in multiple datasets.</p>
<ul>
      <li><a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/linkSpecs/ABA-to-Wiki.silk.xml?rev=176428845b9594e28a2f0362916de23cc821502c">ABA-to-Wiki.silk.xml</a>, </li>
      <li><a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/linkSpecs/KEGG-GENES-to-Wiki.silk.xml?rev=176428845b9594e28a2f0362916de23cc821502c">KEGG-GENES-to-Wiki.silk.xml</a>, </li>
      <li><a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/linkSpecs/KEGG-Pathway-to-Wiki.silk.xml?rev=176428845b9594e28a2f0362916de23cc821502c">KEGG-Pathway-to-Wiki.silk.xml</a>, </li>
      <li><a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/linkSpecs/PharmGKB-to-Wiki.silk.xml?rev=176428845b9594e28a2f0362916de23cc821502c">PharmGKB-to-Wiki.silk.xml</a>, </li>
      <li><a href="http://www.assembla.com/code/ldif/git/nodes/ldif/ldif-singlemachine/src/main/resources/ldif/local/example/test2/linkSpecs/Uniprot-to-Wiki.silk.xml?rev=176428845b9594e28a2f0362916de23cc821502c">Uniprot-to-Wiki.silk.xml</a></li>
    </ul>
<p>To run the example, please download LDIF and use the following LDIF
configuration. The configuration options are explained in the Section <a href="#config">Configuration</a> below. <br>
    </p>
<ul><pre>&lt;LDIF&gt;<br>  &lt;properties&gt;test1k.properties&lt;/properties&gt;<br>  &lt;sources&gt;sources&lt;/sources&gt;<br>  &lt;linkSpecifications&gt;linkSpecs&lt;/linkSpecifications&gt;<br>  &lt;mappings&gt;mappings&lt;/mappings&gt;<br>  &lt;output&gt;output.nq&lt;/output&gt;<br>&lt;/LDIF&gt;</pre>
    </ul>
<p>Execution
      instructions:</p>
<ul>
  <li>Change into the LDIF root
          dir.</li>
  <li>Under Unix type:</li>
  <ul>
    <pre>bin/run examples/1k/config.xml</pre>
    </ul>
  <li>Under
        Windows type:</li>
  <ul>
    <pre>bin\run.bat examples\1k\config.xml</pre>
  </ul>
  </ul>
<h3>Example Data Translation </h3>
<p>In the following, we explain the data translation that is performed
for the example of one entity that is described in two input datasets:</p>
<ul>
  <li><strong>Example <span style="font-style: italic;">input</span> </strong>(reduced to two source datasets, represented using the <a href="http://www4.wiwiss.fu-berlin.de/bizer/TriG/">TriG Syntax</a>):<br>
                      </li>
</ul>
<pre style="margin-left: 40px;">01:  @prefix aba-voc: &lt;http://brain-map.org/gene/0.1#&gt; .<br>02:  @prefix rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt; .<br>03:  @prefix uniprot: &lt;http://purl.uniprot.org/core/&gt; . <br>04: &nbsp;<br>05:  &lt;file:///aba_mouse_20101010_1000.nq&gt; {<br>06:    &lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; aba-voc:entrezgeneid "<span style="font-weight: bold;">18387</span>" ;<br>07:      aba-voc:gene-aliases _:Ab12290 .<br>08:   _:Ab12290 &lt;http://brain-map.org/gene/0.1#aliassymbol&gt; "<span style="font-weight: bold;">Oprk1</span>" .<br>09:  }<br>10:<br>11:  &lt;file:///datasets/uniprot-organism-human-reviewed-complete_1000.nq&gt; {<br>12:    &lt;http://purl.uniprot.org/uniprot/P61981&gt; rdfs:seeAlso &lt;http://purl.uniprot.org/geneid/<span style="font-weight: bold;">18387</span>&gt; .<br>13:    &lt;http://purl.uniprot.org/geneid/<span style="font-weight: bold;">18387</span>&gt; uniprot:database "GeneID" .<br>14:    &lt;http://purl.uniprot.org/uniprot/P61981&gt; uniprot:encodedBy &lt;file:///storage/datasets/uniprot-organism-human-reviewed-complete.rdf#_503237333438003B&gt; .<br>15:  }<br></pre>
<ul>
  <li><strong>Example <span style="font-style: italic;">output</span> :</strong><br>
  </li>
</ul>
<pre style="margin-left: 40px;">01: @prefix smwprop: &lt;http://mywiki/resource/property/&gt; .<br>02: @prefix xsd: &lt;http://www.w3.org/2001/XMLSchema#&gt; .<br>03: <br>04: &lt;file:///aba_mouse_20101010_1000.nq&gt; {<br>05:  &nbsp;&lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; smwprop:EntrezGeneId "<span style="font-weight: bold;">18387</span>"^^xsd:int .<br>06:  &nbsp;&lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; smwprop:GeneSymbol "<span style="font-weight: bold;">Oprk1</span>"^^xsd:string .<br>07: }<br>08: <br>09: &lt;file:///datasets/uniprot-organism-human-reviewed-complete_1000.nq&gt; {<br>10:   &lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; smwprop:EntrezGeneId "<span style="font-weight: bold;">18387</span>"^^xsd:int .<br>11:   &lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; <span style="font-family: monospace;">owl:sameAs </span><span style="font-family: monospace;">&lt;file:///storage/datasets/uniprot-organism-human-reviewed-complete.rdf#_503237333438003B&gt; .</span><br>12: }<br></pre>
<p>The example input and output needs some explanation:</p>
<ul>
        <li>There are two source graphs, each containing data from a
        different source: ABA (input: line 5 to 9) and Uniprot (input: line 11 to 15).</li>
    </ul>
<p>Identity resolution: </p>
<ul>
  <li>Both graphs contain data about the same entity:</li>
  <ul>
        <li>In the ABA dataset the entity is identified using the URI <span style="font-family: monospace;">&lt;http://brain-map.org/mouse/brain/Oprk1.xml&gt; </span>(input: line 6)</li>
        <li>In the Uniprot dataset the entity is identified using the URI <span style="font-family: monospace;"></span><br>
        <span style="font-family: monospace;">&lt;file:///storage/datasets/uniprot-organism-human-reviewed-complete.rdf#_503237333438003B&gt;</span> (input: line 14)</li>
      </ul>
  <li>Since the Silk identity resolution heuristic concludes that both URIs 


 identify

 the same entity, the both URIs are replaced in the output
        with a single URI (in
      this case the ABA one, output: lines 5, 6 and 10).</li>
  <li>The rewritten URI is linked by <span style="font-family: monospace;">owl:sameAs</span> to the original URI (output: line 11).<br>
  </li>

  </ul>
<p>Data Translation: </p>
<ul>
      <li>In the target vocabulary Entrez Gene IDs should be
represented using the smwprop:EntrezGeneId property. Property values
should be represented as xsd:Integers. </li>
      <li>Thus, the
aba-voc:entrezgeneid triple in the first graph (line 6) is translated
into a smwprop:EntrezGeneId triple in the output data (line 5) and a
datatype URI is added to the literal. </li>
      <li>The
smwprop:GeneSymbol triple in line 6 of the output is generated by a
structural transformation out of the two triples in lines 7 and 8 of
the input data. </li>
      <li>In the Uniprot case the <span style="font-family: monospace;">smwprop:EntrezGeneId</span>
      value was extracted from the URI string <span style="font-family: monospace;">&lt;http://purl.uniprot.org/geneid/18387&gt; </span><span style="font-family: sans-serif;">(input: line 12)</span><span style="font-family: monospace;"><span style="font-family: sans-serif;">. </span></span></li>
      <li>The quad with the property <span style="font-family: monospace;">smwprop:EntrezGeneId</span> on line 10 in the output was produced by a complex mapping that had to consider all three quads of the input (lines 12-14).<br>
      </li>
    </ul>
<h2 id="performance"><a name="performance"></a>5.
Performance Evaluation</h2>
<p>We regularly carry out performance evaluations. For more
details and the latest results please visit our <a href="benchmark.html">Benchmark results</a> page.</p>


<h2 id="configuration"><a name="config"></a>6. Configuration
Options</h2>
<p>This section describes how an LDIF configuration file looks like and
which parameters you can modify to change the runtime behavior of LDIF.</p>
<h3>LDIF configuration file</h3>
<p>LDIF is configured with an XML file, whose current structure is very
simple because the integration flow is static at the moment - something
that will change in future releases. <br>
A typical config file looks like
this:<br>
</p>
<pre>&lt;LDIF&gt;<br>  &lt;properties&gt;test.properties&lt;/properties&gt;<br>  &lt;sources&gt;sources&lt;/sources&gt;<br>  &lt;linkSpecifications&gt;linkSpecs&lt;/linkSpecifications&gt;<br>  &lt;mappings&gt;mappings&lt;/mappings&gt;<br>  &lt;output&gt;output.nq&lt;/output&gt;<br>&lt;/LDIF&gt;</pre>
<p>It has the following elements:</p>
<ul>
  <li><span style="font-style: italic;">Properties</span>: The path to
a Java properties file for configuration parameters. See the <a href="#configurationproperties">following&nbsp; section</a> for more details.<br>
</li>
  <li><span style="font-style: italic;">Sources</span>: A directory
containing the source data sets.  These files in this directory must be in <a href="http://sw.deri.org/2008/07/n-quads/">N-Quads</a> format and may
be compressed (.gz, .zip or .bz2).</li>
  <li><span style="font-style: italic;">LinkSpecifications</span>: A
directory containing the <a href="http://www.assembla.com/wiki/show/silk/Link_Specification_Language">Silk
link specifications</a>.</li>
  <li><span style="font-style: italic;">Mappings</span>: A directory
containing the <a href="http://www4.wiwiss.fu-berlin.de/bizer/r2r/spec/#mappings">R2R
mappings</a>.</li>
  <li><span style="font-style: italic;">Output</span>: The name of the file to which the
output should be written.<br>
  </li>
</ul>
<p>All the paths are relative to the config file location. In this case
there is a root directory with the config file and the test.properties
file in it. Furthermore the following directories would be nested in
the root directory: linkSpecs, sources and mappings. As you can see,
currently the datasets have to be in a local directory. More
sophisticated data access will be implemented with the coming <a href="#nextSteps">Data
Access Modules</a>.<br>
</p>
<h3><a name="configurationproperties"></a>Configuration Properties</h3>
<p>In the LDIF configuration file you can specify a (Java) properties
      file to further tweak certain parameters concerning the integration
      workflow. Here is a list with all properties that can be set at the
      moment and the possible values for each property:<br>
      </p>
<ul>
  <li>Specify if all input quads from the input should  be included in the output
        file or only the quads that were mapped/translated by LDIF. Default: <span style="font-family: monospace;">mapped-only</span></li>
</ul>
<pre style="margin-left: 40px;"><span style="font-weight: bold;">output = all | mapped-only</span><br></pre>
<ul>
  <li>Specify if URI aliases in the input data should be rewritten to a single target URI in the output data. Default: <span style="font-family: monospace;">true</span></li>
</ul>
<pre style="margin-left: 40px;"><span style="font-weight: bold;">rewriteURIs = true | false</span><br></pre>
<ul>
  <li>
    <p>Specify the graph containing the provenance information. As of LDIF V0.2, Quads
from this graph are only written to the final output dataset, but are
 processed any further in the integration workflow. Default graph name: <span style="font-family: monospace;">http://www4.wiwiss.fu-berlin.de/ldif/provenance</span></p>
  </li>
</ul>
<pre style="margin-left: 40px;"><span style="font-weight: bold;">provenanceGraphURI = http://www4.wiwiss.fu-berlin.de/ldif/provenance</span><br></pre><ul>
<li>
    <p>Source datasets, R2R mappings and Silk link specifications
are all validated before starting with the actual integration. Since
the syntax validation of the sources (N-Triples / N-Quads files) takes
some time (about 15s/GB), if you already know that they are correct, it
is
possible to disable this step by setting the property to <span style="font-family: monospace;">false</span>. Default:&nbsp;<span style="font-family: monospace;">true</span> <span style="font-family: monospace;"></span></p>
  </li>
</ul>



<pre style="margin-left: 40px;"><span style="font-weight: bold;">validateSources = true | false</span></pre>


<ul>
<li>
    <p>Besides discovering equal entities in the identity
resolution phase, LDIF also offers the opportunity to input these
relationships in form of owl:sameAs links. The NT/N-Quads file with
these sameAs-links has to be placed in the source directory with the
other datasets. If you don't want to use sameAs-links from the input
data, set this property to <span style="font-family: monospace;">false</span>. Default:&nbsp;<span style="font-family: monospace;">true</span> <span style="font-family: monospace;"></span></p>
  </li>
</ul>




<pre style="margin-left: 40px;"><span style="font-weight: bold;">useExternalSameAsLinks = true | false</span><br></pre>
<ul>
<li>
    <p>Although the default output format is N-Quads, LDIF also offers a triple output as N-Triple. Default: <span style="font-family: monospace;">nq</span> <span style="font-family: monospace;"></span></p>
  </li>
</ul>





<pre style="margin-left: 40px;"><span style="font-weight: bold;">outputFormat = nq | nt</span></pre>


<h2 id="development"><a name="code"></a>7. Source
Code and Development</h2>
<p>The latest source code is available from the <a href="http://www.assembla.com/spaces/ldif/wiki">LDIF development page</a> on Assembla.com.</p>
<p>The framework can be used under the terms of the <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache Software
License</a>.
</p>

<h2 id="history"><a name="code"></a>8. Version history </h2>
<p>
</p><table>
    <tbody><tr>
        <th> Version</th>
        <th> Release log</th>
        <th> Date</th>
    </tr>
    <tr class="even">
        <td> 0.2</td>
        <td>
        R2R data translation tasks are now executed in parallel <br>
        Perform source syntax validation before loading data (optional)    <br>
        Support for external sameAs links   <br>
        RDF/N-Triples data output module  <br>
        Support for bzip2 source compression <br>
        Improved loading perfomance <br>
        Memory usage improvements: caching factum rows and string interning only for relevant data
        </td>
        <td> 8/25/2011</td>
    </tr>
    <tr class="odd">
        <td> 0.1</td>
        <td> Intial release of LDIF</td>
        <td> 6/29/2011</td>
    </tr>
</tbody></table>




<h2><a name="feedback" id="feedback"></a>9. Support and Feedback </h2>
<p>For questions and feedback please use the <a href="http://groups.google.com/group/ldif?hl=en">LDIF Google Group</a>. </p>

<h2 id="relatedprojects"><a name="related"></a>10. References</h2>
<ul>
  </ul>
<ul>

  
</ul>
<ol>
  <li>Tom Heath, Christian Bizer: <a href="http://www.morganclaypool.com/doi/abs/10.2200/S00334ED1V01Y201102WBE001">Linked
Data: Evolving the Web into a Global Data Space</a>. Synthesis
Lectures on the Semantic Web: Theory and Technology, Morgan &amp;
Claypool Publishers, ISBN <a href="http://www.amazon.com/Linked-Data-Synthesis-Lectures-Engineering/dp/1608454304/ref=sr_1_1?ie=UTF8&amp;s=books&amp;qid=1297846088&amp;sr=8-1">978160845431</a>,
2011 ( <a href="http://linkeddatabook.com/">Free HTML version</a> ).</li>
  <li>Christian Bizer, Andreas Schultz: <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/BizerSchultz-COLD-R2R-Paper.pdf">The
R2R Framework: Publishing and Discovering Mappings on the Web</a> ( <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/BizerSchultz-COLD-R2R-Talk.pdf">Slides</a>
). 1st International Workshop on Consuming Linked Data (COLD
2010), Shanghai, November 2010. </li>
  <li>Julius Volz, Christian Bizer, Martin Gaedke, Georgi Kobilarov: <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/VolzBizerGaedkeKobilarov-ISWC2009-Silk.pdf">Discovering
and Maintaining Links on the Web of Data</a> ( <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/VolzBizerGaedkeKobilarov-ISWC2009-Silk-Talk.pdf">Slides</a>
). International Semantic Web Conference (ISWC2009),
Westfields, USA, October 2009. </li>

  
  <li>Robert Isele, Anja Jentzsch, Christian Bizer: <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/IseleJentzschBizer-Silk-Cold2010.pdf">Silk
Server - Adding missing Links while consuming Linked Data</a> ( <a href="http://www.wiwiss.fu-berlin.de/en/institute/pwo/bizer/research/publications/IseleJentzschBizer-Silk-Cold2010-Talk.pdf">Slides</a>
). 1st International Workshop on Consuming Linked Data (COLD
2010), Shanghai, November 2010. </li>

</ol>


<h2><a name="acknowledgments"></a>11. Acknowledgments </h2>
<p>This work was supported in part by Vulcan Inc. as part of its <a href="http://www.projecthalo.com">Project Halo</a> and by the EU FP7
project <a href="http://lod2.eu/">LOD2 - Creating Knowledge out of
Interlinked Data</a> (Grant No. 257943). </p>
</div>

</body></html>